src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+
+    void unlink(Node<T> value) {
+        Node<T> next = value.next;
+        Node<T> prev = value.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            value.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            value.next = null;
+        }
+        listSize--;
+    }
+
     @Override
     public void add(T value) {
+        if (listSize == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            first = newNode;
+            last = newNode;
+            first.next = last;
+            last.prev = first;
+        } else {
+            Node<T> newLast = new Node<>(last, value, null);
+            last.next = newLast;
+            last = newLast;
+            first.prev = null;
+        }
+        listSize++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > listSize) {
+            throw new IndexOutOfBoundsException("No such index in the list");
+        }
+
+        if (listSize == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            first = newNode;
+            last = newNode;
+            first.next = last;
+            last.prev = first;
+            listSize++;
+        } else if (index == 0) {
+            Node<T> firstCopy = first;
+            Node<T> newNode = new Node<>(null, value, first);
+            firstCopy.prev = newNode;
+            first = newNode;
+            listSize++;
+        } else if (index == listSize) {
+            Node<T> lastCopy = last;
+            Node<T> newNode = new Node<>(last, value, null);
+            lastCopy.next = newNode;
+            last = newNode;
+            listSize++;
+        } else {
+            addBefore(value, nodeByIndex(index));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public void addAll(List<T> list) {
+        Object[] inputList = list.toArray();
+        int inputLength = inputList.length;
+
+        if (inputLength > 0) {
+            for (Object o : inputList) {
+                T value = (T) o;
+                add(value);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return nodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = nodeByIndex(index);
+        T oldValue = newNode.value;
+        newNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T oldValue = nodeByIndex(index).value;
+        unlink(nodeByIndex(index));
+
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = 0;
+        for (Node<T> x = first; x != null; x = x.next) {
+            if (x.value == null || x.value.equals(object)) {
```

Comment:
```suggestion
            if (x.value == object || x.value != null && x.value.equals(object)) {
```
ця перевірка не працюватиме з налами